### ISSUE TITLE
[FEA] Remove `#include <thrust/optional.h>` from types.hpp

### ISSUE THREAD

**Issue by codereport** [rapidsai/cudf#6475]
Currently, libcudf build fails if you remove the unused:
```cpp
#include <thrust/optional.h>  // TODO no idea why this is needed ¯\_(ツ)_/¯
```
from the file. The build needs to be fixed so that this doesn't happen.

**Comment by codereport** [comment]
This is an issue that still needs to be addressed.

**Comment by vyasr** [comment]
@codereport just so you're aware, I found and resolved this problem while working on something else. Fix is in #7667.
